diff --git a/MODULE.bazel b/MODULE.bazel
index 5ae3ab0..54ca5ec 100644
--- a/MODULE.bazel
+++ b/MODULE.bazel
@@ -16,11 +16,6 @@ bazel_dep(name = "rules_shell", version = "0.3.0")
 bazel_dep(name = "platforms", version = "0.0.11")
 bazel_dep(name = "protobuf", version = "29.0", repo_name = "com_google_protobuf")
 bazel_dep(name = "nlohmann_json", version = "3.12.0.bcr.1", repo_name = "com_github_nlohmann_json")
-bazel_dep(
-    name = "swift_argument_parser",
-    version = "1.3.1.2",
-    repo_name = "com_github_apple_swift_argument_parser",
-)
 
 non_module_deps = use_extension("//swift:extensions.bzl", "non_module_deps")
 use_repo(
